modules/apps/oauth2-provider/oauth2-provider-service/src/main/java/com/liferay/oauth2/provider/service/impl/OAuth2ScopeGrantLocalServiceImpl.java

```diff
@@ -127,9 +127,19 @@ public Collection<OAuth2ScopeGrant> getOAuth2ScopeGrants(
 			return Collections.emptyList();
 		}
 
-		return oAuth2ScopeGrantPersistence.
-			getOAuth2AuthorizationOAuth2ScopeGrants(
+		List<OAuth2ScopeGrant> oAuth2ScopeGrants =
+			oAuth2ScopeGrantPersistence.getOAuth2AuthorizationOAuth2ScopeGrants(
```

Comment:
The companyId parameter is still not used. The underneath OAuth2AuthorizationLocalServiceImpl.fetchOAuth2AuthorizationByAccessTokenContent() is querying with CompanyThreadLocal.getCompanyId(), which could be different from the parameter companyId.

If that happens you would be getting an OAuth2Authorization from different company than what the invoker is asking for, then further derives the other company's OAuth2ScopeGrant for returning.



There is another logical bug here, technically speaking 2 different OAuth2Authorization could have same accessTokenContent. OAuth2AuthorizationLocalServiceImpl.fetchOAuth2AuthorizationByAccessTokenContent() only returns the first match and ignoreing others.
Since OAuth2Authorization and OAuth2ScopeGrant have a many to many relationship, ignoring some OAuth2Authorization could means ignoring many more OAuth2ScopeGrants that are associated to them.



The above 2 issues did not exist in the original 3 table joining solution, changing this to use persistence finder for caching is correct, but it should not alter the original logic..

Reply:
Thanks @shuyangzhou, I completed missed the first issue, and didn't think the second issue would be a problem, but with your explanation I see it can be now.

I've added an additional commit which should resolve both of these issues:
1. Instead of returning just the first OAuth2Authorization, [we return all of them now](https://github.com/liferay-appsec/liferay-portal/pull/2033/commits/1a73aef51758e6c6c4ad871d23fba451919935d1#diff-e2f31c3fdae356032a8891c4fc5fdd914e04dab5792ca50096b387806683b8c2R121-R123).  We then retrieve [all possible OAuth2ScopeGrants based on this list](https://github.com/liferay-appsec/liferay-portal/pull/2033/commits/1a73aef51758e6c6c4ad871d23fba451919935d1#diff-e2f31c3fdae356032a8891c4fc5fdd914e04dab5792ca50096b387806683b8c2R131-R136), and filter as we were before.
2. We now explicitly use [the companyId param](https://github.com/liferay-appsec/liferay-portal/pull/2033/commits/1a73aef51758e6c6c4ad871d23fba451919935d1#diff-e2f31c3fdae356032a8891c4fc5fdd914e04dab5792ca50096b387806683b8c2R123) instead of the CompanyThreadLocal value.  This should also allow us to not have to filter by companyId during OAuth2ScopeGrant processing, since the correct companyId is now implied.

The initial failing tests pass locally with my new changes.  Please let me know if there is anything I've missed, thanks!